### PR TITLE
T22203: Add support for /feed endpoint from libcontentfeed

### PR DESCRIPTION
**Pull request by smspillaz**
https://phabricator.endlessm.com/T22203

### CODE REVIEW

**Comment by cosimoc** [comment]
@smspillaz Please rebase this branch to the latest master.

File: com.endlessm.CompanionAppService.json.in [hunk 1]
```diff
@@ -22,6 +22,7 @@
         "--talk-name=ca.desrt.dconf",
         "--talk-name=com.endlessm.EknServices3.SearchProviderV3",
         "--talk-name=com.endlessm.EknServices2.SearchProviderV2",
+        "--talk-name=com.endlessm.EknServices.SearchProviderV1",
```
Comment (cosimoc) [inline com.endlessm.CompanionAppService.json.in]: This should not be part of this commit (or PR actually)

File: eoscompanion/format.py [hunk 2]
```diff
@@ -37,14 +37,23 @@ def format_app_icon_uri(icon_name, device_uuid):
...
-        contentId=urllib.parse.urlparse(model['thumbnail_uri']).path[1:]
-    ) if model.get('thumbnail_uri', None) else None
+        contentId=urllib.parse.urlparse(thumbnail_uri).path[1:]
+    )
+
+
+def optional_format_thumbnail_uri(application_id, model, device_uuid):
+    '''Format a uri to get an icon for an app if the model has a thumbnail.'''
+    return (
+        format_thumbnail_uri(model['thumbnail_uri'])
```
Comment (cosimoc) [inline eoscompanion/format.py]: You have to pass 3 arguments here; this is fixed later in an unrelated commit, but it should be done here instead.

File: eoscompanion/format.py [hunk 3]
```diff
@@ -50,25 +50,31 @@ def format_thumbnail_uri(application_id, thumbnail_uri, device_uuid):
...
-        format_thumbnail_uri(model['thumbnail_uri'])
+        format_thumbnail_uri(application_id,
+                             model['thumbnail_uri'],
+                             device_uuid)
         if model.get('thumbnail_uri', None)
         else None
     )
 
 
+def format_content_data_uri(content_id, application_id, query):
```
Comment (cosimoc) [inline eoscompanion/format.py]: Why not pass the `device_uuid` directly instead of the whole query?

File: eoscompanion/routes.py [hunk 4]
```diff
@@ -63,6 +62,8 @@
 from .ekn_query import ascertain_application_sets_from_models
 from .format import (
     format_app_icon_uri,
+    format_thumbnail_uri,
```
Comment (cosimoc) [inline eoscompanion/routes.py]: It does not look like this import is needed in this commit

File: eoscompanion/eknservices_bridge.py [hunk 5]
```diff
@@ -164,7 +164,7 @@ def _on_received_feed_query_results(_, result):
             return
 
         callback(None, [
-            ordered.model
+            ordered.get_property('model')
```
Comment (cosimoc) [inline eoscompanion/eknservices_bridge.py]: Looks like this should be squashed with `eknservices: Add method to fetch the content feed from EosDiscoveryFeed`?

File: eoscompanion/routes.py [hunk 6]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+                                    msg,
+                                    path,
+                                    query,
+                                    context,
+                                    content_db_conn):
+    '''Request the Content Feed from EosDiscoveryFeed.
+
+    The :mode: paramter is used to determine whether newer entries
+    should be fetched or older entries should be fetched. It is currently
+    unused for now, the feed is the same as the Discovery Feed on the
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Either remove `currently` or remove `for now` (I'd remove `currently` since you use `for now` below as well)

File: eoscompanion/routes.py [hunk 7]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+        def _on_received_sources(error, sources):
+            '''Callback for when we work out the info for all the sources.'''
+            if error is not None:
+                json_response(msg, {
+                    'status': 'error',
+                    'error': serialize_error_as_json_object(
+                        EosCompanionAppService.error_quark(),
+                        EosCompanionAppService.Error.FAILED,
+                        detail={
+                            'server_error': str(error)
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Did we not say to use `message` for these?

File: eoscompanion/routes.py [hunk 8]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+            server.unpause_message(msg)
+
+        if error is not None:
+            json_response(msg, {
+                'status': 'error',
+                'error': serialize_error_as_json_object(
+                    EosCompanionAppService.error_quark(),
+                    EosCompanionAppService.Error.FAILED,
+                    detail={
+                        'server_error': str(error)
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Ditto

File: eoscompanion/routes.py [hunk 9]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+            ),
+            'contentType': 'text/html'
+        }
+    }]
+
+
+def _serialize_quote_content_feed_model(model):
+    '''Serialize the QUOTE_CARD type content feed model.'''
+    return {
+        'itemType': 'wordOfTheDay',
```
Comment (cosimoc) [inline eoscompanion/routes.py]: `quoteOfTheDay`?

File: eoscompanion/routes.py [hunk 10]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+                model.get_property('thumbnail-uri'),
+                query['deviceUUID']
+            ),
+            'uri': format_content_data_uri(
+                parse_uri_path_basename(model.get_property('uri')),
+                app_id,
+                query
+            ),
+            'duration': model.get_property('duration'),
+            'contentType': 'video/webm'
```
Comment (cosimoc) [inline eoscompanion/routes.py]: Do we not have this as part of the model? It feels fragile to hardcode it here (also for the other types)

File: eoscompanion/routes.py [hunk 11]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+
+    app_infos_for_feed_models(models, _on_received_app_infos)
+
+
+_SOURCE_IDENTIFIER_KEYS_BY_TYPE = {
+    'application': 'applicationId'
+}
+
+
+def get_source_identifier(source):
```
Comment (cosimoc) [inline eoscompanion/routes.py]: This whole machinery feels quite awkward, since there is only one type and one identifier key...

File: eoscompanion/routes.py [hunk 12]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+                app_id,
+                query
+            ),
+            'contentType': 'text/html'
+        }
+    }]
+
+
+def maybe_get_property(model, prop_name):
+    '''Get the year from the first-date property of the model, if it exists.'''
```
Comment (cosimoc) [inline eoscompanion/routes.py]: This is not what this method does

File: eoscompanion/routes.py [hunk 13]
```diff
@@ -161,19 +160,419 @@ def desktop_id_to_app_id(desktop_id):
...
+                logging.error('Encountered error in getting app info: %s', error)
+                continue
+
+            app_infos.append(info)
+
+        callback(None, app_infos)
+
+    def _load_application_info_thunk(desktop_id):
+        '''A thunk used here since we cannot use lambdas in for loops.'''
+        def _internal(callback):
```
Comment (cosimoc) [inline eoscompanion/routes.py]: I'd call this with a different name, since `callback` is already the parameter for the top parent function

File: test/test_service.py [hunk 14]
```diff
@@ -379,14 +386,14 @@ def matches_uri_query(path, expected_query):
...
-FAKE_SHARD_CONTENT = {
+_FAKE_SHARD_CONTENT_PER_APP = {
     'org.test.NoDisplayApp': {
     },
     'org.test.VideoApp': {
         'video_file': {
             'metadata': {
                 'contentType': 'video/mp4',
-                'duration': 304,
+                'duration': '6:04',
```
Comment (cosimoc) [inline test/test_service.py]: This should be moved to the other commit

File: eoscompanion/routes.py (hunk 4 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: Ah, you mean `format_thumbnail_uri` ?

Comment (smspillaz) [reply eoscompanion/routes.py]: Hmm, both `format_content_data_uri`, `format_thumbnail_uri` and `optional_format_thumbnail_uri` all seem to be used below. Not sure what you're referring to here?

File: eoscompanion/routes.py (hunk 11 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: It is, though I'm designing for future sensibility here (we will eventually have card sources that are not necessarily applications and I don't want to start encoding that association now when I know it will have to change later).

File: eoscompanion/routes.py (hunk 10 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: It does feel fragile.

I'm not sure if it is part of the returned discovery feed models, but it certainly exists in the DmContent models. I'll have to add it to libcontentfeed I guess...

File: eoscompanion/routes.py (hunk 6 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: Actually, repeating adjectives is a little mental bug I seem to have. Thanks for catching, I removed the last "for now".

File: eoscompanion/routes.py (hunk 7 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: We did! Legacy strikes again

**Comment by cosimoc** [comment]
(I just pushed a rebased version to review this again)

File: com.endlessm.CompanionAppService.json.in [hunk 15]
```diff
@@ -49,6 +49,17 @@
...
             ]
         },
+        {
+            "name": "libcontentfeed",
+            "buildsystem": "meson",
+            "sources": [
+                {
+                    "type": "git",
+                    "url": "https://github.com/endlessm/libcontentfeed",
+                    "branch": "T22203-content-type-followup"
```
Comment (smspillaz) [inline com.endlessm.CompanionAppService.json.in]: I'll need to fix this

**Comment by smspillaz** [comment]
I rebased and pushed this branch again to remove the dependency on the libcontentfeed branch
